### PR TITLE
rewrite predicates before pushing to union inputs

### LINKED ISSUE #1509: CTE/WITH .. UNION ALL confuses name resolution in WHERE
**Describe the bug**
A construction like:

```sql
with x as (
  select a from baz
  union all
  select a from bar
)
select a 
from x
where a=1
```

..gives the error:

> Plan("No field named 'x.a'. Valid fields are 'baz.a'.")

Explicitly specifying `where x.a` gives the same error, so that is resolving correctly?

Writing the query as `where baz.a=1` instead gives:

> Plan("No field named 'baz.a'. Valid fields are 'x.a'.")

**To Reproduce**

```
❯ create table bar as select 1 as a;
❯ create table baz as select 1 as a;
❯ with x as (select a from baz union all select a from bar) select a from x;
+---+
| a |
+---+
| 1 |
| 1 |
+---+
```

```
❯ with x as (select a from baz union all select a from bar) select a from x where a=1;
Plan("No field named 'x.a'. Valid fields are 'baz.a'.")
```

**Expected behavior**
I believe `SELECT a FROM x WHERE a=1` should function as if x was a single table containing a column named `a`, as shown in non-filtered select in the reproduction steps above.

**Additional context**
This isn't something to do with memory tables; it happens to my Parquet EXTERNAL tables too.

The error messaging here isn't great for my parquet files, which already use dotted notation, like `select "address.line1" from address`, now we have "address.address.line1".

### PR DESCRIPTION
**Pull request by korowa**
# Which issue does this PR close?

Closes #1509.

 # Rationale for this change

# What changes are included in this PR?

Currently pushing predicates to union all inputs retains subquery or CTE alias in predicates and generates following plan:
```
Projection: #x.a
  Union
    Projection: #baz.a
      Filter: #x.a = Int64(1)
        TableScan: baz projection=Some([0])
    Projection: #bar.a
      Filter: #x.a = Int64(1)
        TableScan: bar projection=Some([0])
```

which breaks on trying to filter bar/baz tables on x.a field.

Changes included in this PR

1. remove qualifiers from predicates before pushing them down to union inputs
2. modifiy projection plan step to make it able to resolve unqualified column identifiers in filters (this could happen due to predicate being rewritten in union step)

# Are there any user-facing changes?

Fixed execution of queries with filters on CTE/subqueries containing UNION clause

### CODE REVIEW

**Review by alamb — COMMENTED** [review]
Thank you @korowa  -- I will review this later today

File: datafusion/src/optimizer/filter_push_down.rs
Comment (alamb) [inline datafusion/src/optimizer/filter_push_down.rs:360]: Why is this change to `Projection` needed? I tried removing this change locally and all the tests still passed 🤔 

Can you add a test case that would fail without this change?

**Review by alamb — COMMENTED** [review]
Thank you for the contribution @korowa ❤️ 

I have one question related to the change to the `Projection` logic  but otherwise this PR looks good 👨‍🍳 💋 

cc @FauxFaux

File: datafusion/src/optimizer/filter_push_down.rs
Comment (korowa) [reply datafusion/src/optimizer/filter_push_down.rs:360]: Yes, that change is not really necessary, i guess - all downstream operations manages to resolve proper objects for forwarding filters (at least I checked both simple projections and joins). It just looked more consistent to support unqualified columns resolution.

I can revert this change if other parts look fine, np.

Comment (korowa) [reply datafusion/src/optimizer/filter_push_down.rs]: Done, waiting for CI

Thank you!

**Review by alamb — APPROVED** [review]
Thanks @korowa

**Comment by alamb** [comment]
I'll merge this in when the CI checks have passed

**Comment by alamb** [comment]
Thanks again @korowa

**Comment by xudong963** [comment]
Nice work @korowa 👍
I was just wondering if the `INTERSECT` and `EXCEPT` have the same problem. Suddenly, I realized that we implemented both operations using join, lol

**Pull request merged**
